runtime/callback.c

```diff
@@ -240,7 +343,7 @@ CAMLexport void caml_iterate_named_values(caml_named_action f)
   for(i = 0; i < Named_value_size; i++){
     struct named_value * nv;
     for (nv = named_value_table[i]; nv != NULL; nv = nv->next) {
-      f( &nv->val, nv->name );
+      f( Op_val(nv->val), nv->name );
```

Comment:
This looks suspicious. `Op_val` is 

```c
#define Op_val(x) ((value *) (x))
```
Should this be the same as trunk?

Reply:
Agree, this looks wrong, we should revert. 

I see another dubious thing: we don't lock `named_value_lock` in `caml_iterate_named_values`. Is this intentional? Even if intentional, will it still be safe on ARM where the write to the `named_value_table[h] = nv` might happen in an alternate order to the preceding setup of `nv`?

Reply:
Not having a lock seems like an omission. We don't have a test for `caml_iterate_named_values` in the testsuite, which we should.